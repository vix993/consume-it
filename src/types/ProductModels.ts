export interface ProductsData {
    available: number;
    id: number;
    name: string;
    price: number;
}

export interface OrdersData {
    id: number;
    name: string;
    price: number;
    quantity: number;
}

export interface VouchersData {
    id: number;
    code: string;
    type: 'percentual' | 'fixed' | 'shipping';
    amount: number;
    minValue?: number;
}