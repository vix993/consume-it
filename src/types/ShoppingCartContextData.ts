import { ProductsData, OrdersData, VouchersData } from "./ProductModels";

export interface ShoppingCartContextData {
    products: ProductsData[];
    orders: OrdersData[];
    total: number;
    discount: number;
    subtotal: number;
    shipping: number;
    vouchers: VouchersData[];
    activeVoucher: VouchersData | undefined;
    buyProduct: (productId: number) => void;
    removeOrder: (orderProductId: number, quantity: number) => void;
    handleVoucherSelection: (discountCode: string) => string;
}