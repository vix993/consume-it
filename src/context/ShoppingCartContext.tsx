import { createContext, useState, ReactNode, useEffect } from 'react';

import { getProducts } from '../repositories/productsRepository';
import { getVouchers } from '../repositories/vouchersRepository';

import { ProductsData, VouchersData, OrdersData } from '../types/ProductModels';
import { ShoppingCartContextData } from '../types/ShoppingCartContextData';

import { createNewOrdersList } from '../utils/shoppingCartUtils';

export interface ShoppingCartProviderProps {
    children: ReactNode;
}

export const ShoppingCartContext = createContext({} as ShoppingCartContextData)

export const ShoppingCartProvider = ({ children }: ShoppingCartProviderProps) => {
    const [products, setProducts] = useState<ProductsData[]>([]);
    const [orders, setOrders] = useState<OrdersData[]>([]);
    const [vouchers, setVouchers] = useState<VouchersData[]>([]);
    const [activeVoucher, setActiveVoucher] = useState<VouchersData>();
    const [weightOfPurchase, setWeightOfPurchase] = useState(0);
    const [subtotal, setSubtotal] = useState(0);
    const [shipping, setShipping] = useState(0);
    const [shippingOffset, setShippingOffset] = useState(0);
    const [discount, setDiscount] = useState(0);
    const [total, setTotal] = useState(0);

    // fetching product data
    const requestProducts = async () => {
        await getProducts()
            .then((res: any) => {
                if (!res.error)
                    setProducts(res);
            })
            .catch((err: any) => {
                console.log(err);
            });
        // console.log(products);
    }

    const requestVouchers = async () => {
        await getVouchers()
            .then(async (res: any) => {
                console.log(res)
                if (!res.error){
                    setVouchers(res);
                }
            })
            .catch((err: any) => {
                console.log(err);
            })
        // console.log('vouchers', vouchers);
    }

    useEffect(() => {
        requestProducts();
        requestVouchers();
    }, []);

    const updateWeightOfPurchase = () => {
        let newWeight = 0;
        orders.map((order) => {newWeight += order.quantity});
        // console.log(newWeight)
        setWeightOfPurchase(newWeight);
    }

    const updateShippingPrice = () => {
        console.log('hello')
        if (subtotal > 400
            || activeVoucher
            && (activeVoucher.type == 'shipping'
            && subtotal > activeVoucher.minValue!)
            ) {
            setDiscount(shipping)
            console.log('discount in updateshipping', discount);
            // setShipping(0);
            return
        }
        if (weightOfPurchase === 0){
            setShipping(0);
            return
        }

        if (weightOfPurchase === 1)
            setShipping(30);
        if (weightOfPurchase % 5 === 0 || weightOfPurchase % 5 === 4) {
            const baseShipping = weightOfPurchase === 0 ? 0 : 30;
            if (weightOfPurchase < 14){
                const newOffset = (((weightOfPurchase - (weightOfPurchase % 5)) / 5) * 7);
                setShippingOffset(newOffset);
            }
            
            const addedShipping = (((weightOfPurchase - (weightOfPurchase % 5)) * 7) / 5) - shippingOffset
            setShipping(baseShipping + addedShipping)
        }
    }
    console.log(activeVoucher)

    const updateSubtotal = () => {
        let newSubtotal = 0;
        orders.map((order) => {
            newSubtotal += order.price * order.quantity
        })
        if (activeVoucher && activeVoucher.type === 'percentual') {
            setDiscount(newSubtotal - (newSubtotal * (activeVoucher.amount / 100)));
            console.log('discount in updatesubtotal', discount);
        }
        setSubtotal(newSubtotal);
    }

    const updateTotal = () => {
        if (activeVoucher && activeVoucher.type === 'fixed') {
            setDiscount(activeVoucher.amount);
        }
        const newTotal = (subtotal + shipping) - discount;
        setTotal(newTotal);
    }

    const updateOrders = (product: ProductsData) => {
        const newProductPurchased = orders.filter(order => order.id === product.id);
        if (newProductPurchased[0]) {
            setOrders(createNewOrdersList(product, orders, 'add'));
        }
        else {
            setOrders([...orders, {
                id: product.id,
                name: product.name,
                price: product.price,
                quantity: 1
            }])
        }
    }

    const handleVoucherSelection = (discountCode: string) => {
        const voucherSelection = vouchers.filter((voucher) => voucher.code === discountCode)[0];
        if (!voucherSelection)
            return "code invalid";
        console.log(activeVoucher === voucherSelection, voucherSelection);
        if (activeVoucher)
            return "code already active";
        setActiveVoucher(voucherSelection)
        return "discount applied"
    }

    const buyProduct = (productId: number) => {
        let product = products.filter((product) => product.id === productId)[0]
        
        if (product.available > 0) {
            const newProductsData = products.filter((prod: ProductsData) => product.id != prod.id);
            product.available--;
            const newProducts = [...newProductsData, product];
            updateOrders(product);
            setProducts(newProducts.sort((prod, prod1) => prod.id - prod1.id));
            updateWeightOfPurchase();
        }
    }

    const removeOrder = (orderProductId: number, quantity: number) => {
        const changedOrderProduct = products.filter((product) => product.id === orderProductId)[0];
        if (quantity > 0)
            setOrders(createNewOrdersList(changedOrderProduct, orders, 'subtract'));

        const unchangedProducts = products.filter((product) => product.id !== orderProductId);
        const changedProduct = {...changedOrderProduct, available: changedOrderProduct.available + 1}
        const newProducts = [...unchangedProducts, changedProduct].sort((prod, prod1) => prod.id - prod1.id)
        setProducts(newProducts)
    }

    // update weight everytime order changes
    useEffect(() => {
        updateWeightOfPurchase();
        updateSubtotal();
    }, [orders, products, discount]);

    useEffect(() => {
        updateShippingPrice();
    }, [weightOfPurchase, discount, shippingOffset, removeOrder])

    useEffect(() => {
        updateTotal();
    }, [subtotal, shipping, discount, activeVoucher]);

    return (
        <ShoppingCartContext.Provider
            value={{
                products,
                orders,
                subtotal,
                discount,
                total,
                shipping,
                vouchers,
                activeVoucher,
                buyProduct,
                removeOrder,
                handleVoucherSelection,
                updateShippingPrice
            }}
        >
            {children}
        </ShoppingCartContext.Provider>
    );
}