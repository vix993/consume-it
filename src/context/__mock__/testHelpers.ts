import { ShoppingCartContextData } from '../../types/ShoppingCartContextData';
import {data} from  './dummy';

export const context: ShoppingCartContextData = {
    handleVoucherSelection: jest.fn(),
    products: data,
    orders: [
        {id: 1, name: 'Banana', price: 10, quantity:2},
        {id: 2, name: 'Apple', price: 10, quantity:2},
        {id: 3, name: 'Mango', price: 10, quantity:2},
        {id: 4, name: 'Pear', price: 10, quantity:2}
    ],
    total: 0,
    discount: 0,
    subtotal: 0,
    shipping: 0,
    vouchers: [{
        id:	3,
        code: "#SHIPIT",
        type: "shipping",
        amount: 0,
        minValue: 300.5
    }],
    activeVoucher: {
        id:	3,
        code: "#SHIPIT",
        type: "shipping",
        amount: 0,
        minValue: 300.5
    },
    buyProduct: jest.fn().mockImplementation((number) => {}),
    removeOrder: jest.fn().mockImplementation((number, num) => {}),
    updateShippingPrice: jest.fn(),
}