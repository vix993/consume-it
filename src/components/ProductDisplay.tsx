import React, {useContext} from 'react';

import { ShoppingCartContext } from '../context/ShoppingCartContext';

import { Product } from './Product';

import styles from '../styles/components/product-display.module.css';

interface ProductDisplayProps {

}

export const ProductDisplay: React.FC<ProductDisplayProps> = ({}) => {
    const { products } = useContext(ShoppingCartContext);
    console.log(products)
    return (
        <section>
            <div className={styles.product_display_wrapper}>
                {products.map((product) => {
                    console.log(product.name)
                    return (
                        <Product
                            key={product.id}
                            name={product.name}
                            price={product.price}
                            available={product.available}
                        />
                    );
                })}
            </div>
        </section>
    );
}