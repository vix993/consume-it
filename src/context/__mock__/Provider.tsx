import {context} from './testHelpers';

export const MyContext = ({
    Consumer(props: any) {
        return props.children(context)
    }
})