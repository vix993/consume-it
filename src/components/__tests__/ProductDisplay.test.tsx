import React from "react";

import { cleanup, fireEvent, queryByDisplayValue, render, screen } from '@testing-library/react';

import { ProductDisplay } from '../ProductDisplay';

import { ShoppingCartContext } from '../../context/ShoppingCartContext';


import { ShoppingCartContextData } from "../../types/ShoppingCartContextData";

import { context } from "../../context/__mock__/testHelpers";

const renderProductDisplay = (values: ShoppingCartContextData) => {
    return render(
        <ShoppingCartContext.Provider value={values}>
            <ProductDisplay />
        </ShoppingCartContext.Provider>
    )
}

describe("<ProductDisplay />", () => {
    afterEach(cleanup);

    it("Should render a product list with context provided values", () => {
        const {container }= renderProductDisplay(context)
        expect(container).toMatchSnapshot();
        expect(screen.getByText('Banana'))
    });
})