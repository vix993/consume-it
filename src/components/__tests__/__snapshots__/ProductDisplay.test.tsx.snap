// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDisplay /> Should render a product list with context provided values 1`] = `
<div>
  <section>
    <div>
      <section>
        <div />
        <main>
          <span>
            Banana
          </span>
          <div>
            <div>
              $ 
              10
              ,00
            </div>
            <div>
              10
               left
            </div>
          </div>
        </main>
        <div>
          <button>
            BUY
          </button>
        </div>
      </section>
      <section>
        <div />
        <main>
          <span>
            Apple
          </span>
          <div>
            <div>
              $ 
              20
              ,00
            </div>
            <div>
              15
               left
            </div>
          </div>
        </main>
        <div>
          <button>
            BUY
          </button>
        </div>
      </section>
      <section>
        <div />
        <main>
          <span>
            Orange
          </span>
          <div>
            <div>
              $ 
              30
              ,00
            </div>
            <div>
              8
               left
            </div>
          </div>
        </main>
        <div>
          <button>
            BUY
          </button>
        </div>
      </section>
      <section>
        <div />
        <main>
          <span>
            Mango
          </span>
          <div>
            <div>
              $ 
              15
              ,00
            </div>
            <div>
              20
               left
            </div>
          </div>
        </main>
        <div>
          <button>
            BUY
          </button>
        </div>
      </section>
      <section>
        <div />
        <main>
          <span>
            Dragon Fruit
          </span>
          <div>
            <div>
              $ 
              40
              ,00
            </div>
            <div>
              50
               left
            </div>
          </div>
        </main>
        <div>
          <button>
            BUY
          </button>
        </div>
      </section>
      <section>
        <div />
        <main>
          <span>
            Passion Fruit
          </span>
          <div>
            <div>
              $ 
              35
              ,00
            </div>
            <div>
              70
               left
            </div>
          </div>
        </main>
        <div>
          <button>
            BUY
          </button>
        </div>
      </section>
    </div>
  </section>
</div>
`;
